### PR TITLE
[Merged by Bors] - Implement `Promise.all`

### PR DESCRIPTION
**Pull request by raskad**
This Pull Request changes the following:

- Implement `Promise.all`

### CODE REVIEW

**Review by HalidOdat — APPROVED** [review]
Looks perfect to me! :)

File: boa_engine/src/builtins/promise/mod.rs [hunk 1]
```diff
@@ -316,6 +317,247 @@ impl Promise {
...
+            capability_resolve: JsFunction,
+            #[unsafe_ignore_trace]
+            remaining_elements_count: Rc<Cell<i32>>,
+        }
+
+        // 1. Let values be a new empty List.
+        let values = GcCell::new(Vec::new());
+
+        // 2. Let remainingElementsCount be the Record { [[Value]]: 1 }.
+        let remaining_elements_count = Rc::new(Cell::new(1));
```
Comment (Razican) [inline boa_engine/src/builtins/promise/mod.rs:410]: As far as I can tell, the reason behind this being inside cells is that we need to pass them as captures below, right? But we don't use them until step `j`, if I'm not mistaken. Wouldn't it be more efficient to wrap them in cells after the `loop` has completed?

**Review by Razican — APPROVED** [review]
Looks good, thank you!
I'm happy to merge it from my side, but I'd like to point out a potential performance optimization. Give it a look, and if it makes sense, feel free to implement it and merge.

**Pull request closed without merging**